File: framework/src/datamirai_engine/tools/builtin/filesystem/file_info.py
"""filesystem/file_info — get file or directory metadata."""

from __future__ import annotations

import os
import stat as _stat
from typing import Any

from datamirai_engine.core.context import ExecutionContext
from datamirai_engine.tools.base import BaseTool, ToolInput, ToolOutput, ToolSpec


class FileInfoTool(BaseTool):
    spec = ToolSpec(
        tool_type="filesystem/file_info",
        version="1.0.0",
        display_name="File Info",
        description="Returns metadata about a file or directory: size, permissions, modification time, type.",
        category="filesystem",
        icon="info",
        intents=[
            "obtener informacion de un archivo",
            "verificar si un archivo existe",
            "ver permisos y tamano de un archivo",
        ],
        inputs=[
            ToolInput(name="path", type="string", required=True, description="Path to inspect"),
        ],
        outputs=[
            ToolOutput(name="exists", type="boolean", description="Whether the path exists"),
            ToolOutput(name="type", type="string", description="'file', 'directory', 'symlink', or 'other'"),
            ToolOutput(name="size", type="number", description="Size in bytes"),
            ToolOutput(name="modified", type="number", description="Last modification timestamp"),
            ToolOutput(name="created", type="number", description="Creation timestamp"),
            ToolOutput(name="permissions", type="string", description="Unix permission string (e.g. 'rwxr-xr-x')"),
            ToolOutput(name="path", type="string", description="Resolved absolute path"),
        ],
    )

    async def execute(
        self,
        inputs: dict[str, Any],
        config: dict[str, Any],
        context: ExecutionContext | None,
    ) -> dict[str, Any]:
        path = os.path.abspath(inputs["path"])

        if not os.path.exists(path):
            return {
                "exists": False,
                "type": "unknown",
                "size": 0,
                "modified": 0,
                "created": 0,
                "permissions": "",
                "path": path,
            }

        st = os.stat(path)
        mode = st.st_mode

        if os.path.islink(path):
            ftype = "symlink"
        elif os.path.isdir(path):
            ftype = "directory"
        elif os.path.isfile(path):
            ftype = "file"
        else:
            ftype = "other"

        perm = _stat.filemode(mode)[1:]  # strip leading 'd' or '-'

        return {
            "exists": True,
            "type": ftype,
            "size": st.st_size,
            "modified": st.st_mtime,
            "created": getattr(st, "st_birthtime", st.st_ctime),
            "permissions": perm,
            "path": path,
        }

File: engine/src/tools/builtin/filesystem/file_info.rs
//! FileInfoTool

use super::*;

// ===========================================================================
// FileInfoTool
// ===========================================================================

fs_tool! {
    struct FileInfoTool, factory FileInfoFactory;
    tool_type = "filesystem/file_info",
    name = "File Info",
    description = "Returns metadata about a file: size, modified time, type, permissions",
    category = "filesystem",
    inputs = [
        field("path", FieldType::String, true, "Path to get info for"),
    ],
    outputs = [
        field("path", FieldType::String, true, "Resolved absolute path"),
        field("size", FieldType::Number, true, "Size in bytes"),
        field("modified", FieldType::Number, true, "Last modified timestamp (seconds since epoch)"),
        field("is_file", FieldType::Boolean, true, "Whether path is a file"),
        field("is_dir", FieldType::Boolean, true, "Whether path is a directory"),
        field("permissions", FieldType::String, true, "Permission mode (octal on unix)"),
    ],
    config_fields = []
}

#[async_trait]
impl Tool for FileInfoTool {
    async fn execute(
        &self,
        inputs: HashMap<String, Value>,
        _config: &HashMap<String, Value>,
        _context: &dyn ExecutionContext,
    ) -> Result<HashMap<String, Value>, ToolError> {
        let raw_path = inputs.get("path").and_then(|v| v.as_str()).ok_or_else(|| {
            ToolError::ExecutionFailed {
                tool_type: "filesystem/file_info".into(),
                message: "missing required input: path".into(),
            }
        })?;

        let path = fs::canonicalize(raw_path).map_err(|e| ToolError::ExecutionFailed {
            tool_type: "filesystem/file_info".into(),
            message: format!("Path not found: {raw_path} ({e})"),
        })?;

        let metadata = fs::metadata(&path).map_err(|e| ToolError::ExecutionFailed {
            tool_type: "filesystem/file_info".into(),
            message: format!("Cannot stat path: {e}"),
        })?;

        let modified = metadata
            .modified()
            .ok()
            .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
            .map(|d| d.as_secs_f64())
            .unwrap_or(0.0);

        #[cfg(unix)]
        let permissions = {
            use std::os::unix::fs::PermissionsExt;
            format!("{:o}", metadata.permissions().mode())
        };
        #[cfg(not(unix))]
        let permissions = if metadata.permissions().readonly() {
            "readonly".to_string()
        } else {
            "read-write".to_string()
        };

        let mut out = HashMap::new();
        out.insert("path".to_string(), json!(path.display().to_string()));
        out.insert("size".to_string(), json!(metadata.len()));
        out.insert("modified".to_string(), json!(modified));
        out.insert("is_file".to_string(), json!(metadata.is_file()));
        out.insert("is_dir".to_string(), json!(metadata.is_dir()));
        out.insert("permissions".to_string(), json!(permissions));
        Ok(out)
    }
}

// ---------------------------------------------------------------------------
// Registration helper
